Balsara Method for ComputeEMF

lizmcole opened the pull request:
### Description
Uses the method from Balsara's 2025 paper to compute the emf by getting v x B at cell center, then reconstructing face-centered B-fields to the edge, and then using equations 3.3, 3.5, 3.9, and 3.10 to get the EMF at the edge.

### Related issues
This still has the GPU problem with MHD. e.g.,
https://github.com/quokka-astro/quokka/issues/1258
https://github.com/quokka-astro/quokka/issues/1266

### Checklist
_Before this pull request can be reviewed, all of these tasks should be completed. Denote completed tasks with an `x` inside the square brackets `[ ]` in the Markdown source below:_
- [ x] I have added a description (see above).
- [x ] I have added a link to any related issues (if applicable; see above).
- [x ] I have read the [Contributing Guide](https://github.com/quokka-astro/quokka/blob/development/CONTRIBUTING.md).
- [ ] I have added tests for any new physics that this PR adds to the code.
- [ ] *(For quokka-astro org members)* I have manually triggered the GPU tests with the magic comment `/azp run`.

lizmcole committed:
Adding Balsara version, error  in grid sizes with ec_fabs_Bi_ieside

lizmcole committed:
Added Balsara 2nd order method

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

lizmcole committed:
Merge branch 'development' into Balsara-computeemf

markkrumholz commented on src/hydro/mhd_system.hpp:
```diff
@@ -158,7 +163,7 @@ void MHDSystem<problem_t>::ComputeEMF(std::array<amrex::MultiFab, AMREX_SPACEDIM
 				const amrex::Box box_fc_U = amrex::grow(box_fc, (nghost_cc - 1) * vec_fc2ec + 1);
 
-				// extrapolate both required cell-centered velocity fields to the cell-edge
+				// extrapolate cell-centered electic fields to the cell-edge
```
Typo -- "electic"

markkrumholz commented on src/hydro/mhd_system.hpp:
```diff
@@ -329,6 +334,209 @@ void MHDSystem<problem_t>::ComputeEMF(std::array<amrex::MultiFab, AMREX_SPACEDIM
+		std::array<amrex::FArrayBox, 3> cc_fabs_EMF;
+		{
+			const amrex::Box &box_cc_EMF = amrex::grow(box_cc, nghost_cc); // only need cc ghost zones for Balsara method
+			cc_fabs_EMF[0].resize(box_cc_EMF, 1);
```
See the discussion of the problems of using resize as part of #1263. This should be rewritten using the fix proposed there.

lizmcole replied:
Okay, I copied the original compute-emf function from this and changed the parts of the computeEMF_Balsara that seemed to reflect what needed changed.

markkrumholz replied:
Understood. We are in the process of fixing that, and we will need to apply the same fixes here for this to run stably on GPU.

markkrumholz commented on src/hydro/mhd_system.hpp:
```diff
@@ -329,6 +334,209 @@ void MHDSystem<problem_t>::ComputeEMF(std::array<amrex::MultiFab, AMREX_SPACEDIM
+					const double S1_m = std::max(fspd_x1(i, j, k, 0), fspd_x1(i + delta_w1[0], j + delta_w1[1], k + delta_w1[2], 0));
+					const double S1_p = std::max(fspd_x1(i, j, k, 1), fspd_x1(i + delta_w1[0], j + delta_w1[1], k + delta_w1[2], 1));
+
+					if (i == 0 && j > 7 && k == 0) {
```
Debugging prints should be removed before merging.

lizmcole replied:
removed

markkrumholz replied:
I still see it. Did you perhaps forget to push the update?

lizmcole replied:
I pushed to the wrong branch, should be there now.

markkrumholz commented on src/hydro/mhd_system.hpp:
```diff
@@ -329,6 +334,209 @@ void MHDSystem<problem_t>::ComputeEMF(std::array<amrex::MultiFab, AMREX_SPACEDIM
+					//   LD  |  RD
+					// 		 |
+
+					const auto E2_LD = cc_E2_array(i, j, k);
```
It seems like a lot of the quantities you define here never get used. For example, `B0_dstar` is defined, but then never referenced. Can you go through here and check that you are only defining variables that you are actually going to use? You can also increase efficiency by moving a bunch of these calculations inside the final if block, so for example you don't compute `E2_L_star` in cases where the wavespeeds imply that you don't need it.

lizmcole replied:
Still todo; it was written exactly as it was in the paper so I can double check equations, but yes, some of the star quantities are written but not used and there is definitely a more efficient way to do it.

markkrumholz replied:
Understood. Just ping me when this is ready for re-review.

markkrumholz reviewed (changes requested):
I left comments inline in the code.

BenWibking reviewed (commented):
I don't see any hooks to use the new method - is that the case? I think it would be useful to have one of the test problems use the new method when running in the test suite.

lizmcole committed:
Fixed clangtidy and made changes like #1263

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

lizmcole committed:
Merge branch 'development' into Balsara-computeemf

lizmcole committed:
Added hook in parmparse for Balsara method

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

lizmcole committed:
rollback of QuokkaSimulation, bugs introduced in previous commit

lizmcole committed:
Added parmparse mhd.use_balsara_compute_emf= true/false for input file use

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

lizmcole commented:
The indexing is not correct for E2_LD, LU, RD, and RU. I have to work out how to shift the 4 values thing one box down and to the left relative to the direction for a particular edge.

BenWibking commented on src/hydro/mhd_system.hpp:
```diff
@@ -329,6 +330,204 @@ void MHDSystem<problem_t>::ComputeEMF(std::array<amrex::MultiFab, AMREX_SPACEDIM
+				const auto &fspd_x1 = fcx_mf_fspds[w1_comp][mfi].const_array();
+				const auto &emf_cc = fcx_mf_fspds[w1_comp][mfi].const_array();
+
+				// LLF variant
```
I'm confused by this comment, since I don't think this is actually the LLF variant. From reading Balsara et al. (2025), that appears to be given by Equation 3.9, where there are no special cases for supersonic states and there is one one case to handle. However, 3.10 handles all the varying cases of supersonic states in all the combinations of directions, but that is not an LLF solver and I don't think uses 3.9 at all.

lizmcole replied:
I switched it all over to handle supersonic states now.

lizmcole committed:
Merge branch 'development' into Balsara-computeemf

lizmcole committed:
Fixed indexing

lizmcole committed:
Merge branch 'Balsara-computeemf' of https://github.com/quokka-astro/quokka into Balsara-computeemf
Merge remote changes from 'Balsara-computeemf' to sync updates

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

lizmcole committed:
Adding extrapolation of EMF to cell edges

lizmcole committed:
WIP: local changes before merge

lizmcole committed:
Adding extrapolation of EMF to edges

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

lizmcole committed:
Fixed reconstruct of emf cc -> fc -> ec for Ez component only

lizmcole committed:
Fixed reconstruct of emf cc -> fc -> ec for Ez component only

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

lizmcole committed:
Fixed reconstruct of emf cc -> fc -> ec for Ez component only pt 2

lizmcole committed:
Fixed reconstruct of emf cc -> fc -> ec for Ez component only pt 2

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

lizmcole committed:
Fixed reconstruct of emf cc -> fc -> ec for Ez component only pt 3

lizmcole committed:
Merge branch 'Balsara-computeemf' of https://github.com/quokka-astro/quokka into Balsara-computeemf

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

lizmcole committed:
Using Eqn 3.2, 3.4, 3.6, 3.7, 3.8 from Balsara 2025

lizmcole committed:
Merge branch 'Balsara-computeemf' of https://github.com/quokka-astro/quokka into Balsara-computeemf

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

lizmcole committed:
Removed unused quantities and comments

lizmcole committed:
merged

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

lizmcole committed:
Removed unused parameter

lizmcole committed:
Renamed box_cc_EMF -> box_cc_EMF_1D when used in scope making quadrants

lizmcole committed:
Renamed box_cc_EMF -> box_cc_EMF_edge (more descriptive)  when used in scope making quadrants

BenWibking commented:
/azp run

BenWibking reviewed (changes requested):
It looks like there are a couple of unused variable warnings left:
```
  425 |                                 const auto cc_E2_array = cc_a4_EMF_array[iedge];
      |                                            ^~~~~~~~~~~
/home/runner/work/quokka/quokka/src/hydro/mhd_system.hpp:460:43: error: unused variable ‘w2_comp’ [-Werror=unused-variable]
  460 |                                 int const w2_comp = iedge;
      |                                           ^~~~~~~
```

The GPU failures look like they're unrelated to this PR, but I think Chongchong should figure out what's going on before we merge any new PRs, just in case there's a subtle failure hiding underneath.

Can you point out where in Balsara's paper the reconstruction is explained? Does Balsara do the same reconstruction in both ways to the edge?

BenWibking commented:
Also: is this method stable with the non-EP reconstruction schemes? Whichever way it turns out, that would be extremely helpful to document in this PR.

lizmcole committed:
Removed unused variables

lizmcole commented:
@BenWibking 
Balsara just said to do the EMF reconstruction the same way we already do our other reconstructions when I asked.

BenWibking commented:
> @BenWibking Balsara just said to do the EMF reconstruction the same way we already do our other reconstructions when I asked.

Ok, that's interesting. I think this means there are 4 possible MHD schemes to consider:
1. Felker & Stone reconstruction, LD04 EMF Riemann solver
2. Felker & Stone reconstruction, Balsara EMF Riemann solver
3. Face velocity reconstruction, LD04 EMF Riemann solver
4. Face velocity reconstruction, Balsara EMF Riemann solver (would have to compute EMF in quadrants after velocity reconstruction for this case, rather than cell-centered EMF)

Would it make sense to split the reconstruction and EMF Riemann solver into separate functions in MHDSystem to test these four combinations?

BenWibking commented:
Although the first three are already implemented, so maybe that doesn't actually save any effort. Case 4 seems interesting, and could just be copy-and-pasted into its own ComputeEMF_* function.

Edit: I think the reason case 4 is interesting is that it will let us test the non-EP reconstructions with the Balsara EMF Riemann solver.

markkrumholz commented:
I don't think case 4 exists, because if you're using the Balsara scheme you don't need the velocities at the faces or corners at all. The whole point of the scheme is that you compute the v x B terms at cell centers and then extrapolate that quantity to the corners. When you do that, the velocity by itself never appears at the corners.

BenWibking commented:
> I don't think case 4 exists, because if you're using the Balsara scheme you don't need the velocities at the faces or corners at all. The whole point of the scheme is that you compute the v x B terms at cell centers and then extrapolate that quantity to the corners. When you do that, the velocity by itself never appears at the corners.

I agree that is a point of the scheme, but to me the more salient feature is that Balsara's Riemann solver treats supersonic states whereas the LD04 Riemann solver assumes you are always in the subsonic strongly interacting state. So for low order reconstructions, the Balsara Riemann solver should be significantly less numerically resistive.

markkrumholz commented:
> > I don't think case 4 exists, because if you're using the Balsara scheme you don't need the velocities at the faces or corners at all. The whole point of the scheme is that you compute the v x B terms at cell centers and then extrapolate that quantity to the corners. When you do that, the velocity by itself never appears at the corners.
> 
> I agree that is a point of the scheme, but to me the more salient feature is that Balsara's Riemann solver treats supersonic states whereas the LD04 Riemann solver assumes you are always in the subsonic strongly interacting state. So for low order reconstructions, the Balsara Riemann solver should be significantly less numerically resistive.

OK, that's fine, but in that case I think your taxonomy of possible combinations is a bit confusing. It seems like there are two independent axes that could be varied. One is how you reconstruct the corner EMFs, for which we have three possibilities:

1. Felker & Stone scheme as originally published
2. Our variant of Felker & Stone with velocities coming from the Riemann solver.
3. The Balsara scheme where you compute v x B at cell centers and then extrapolate that to corners.

Then we have two distinct Riemann solvers: the LD04 one and the Balsara one. So I guess in principle there are 6 distinct combinations. I don't know that it makes sense to test all of them, but from a code legibility standpoint we should probably separate out the flags that control the choice of reconstruction to the corner scheme and those that control the corner Riemann solver, since in principle you can (I think) mix and match any of these, though of course not all are guaranteed to work.

BenWibking commented:
> > > I don't think case 4 exists, because if you're using the Balsara scheme you don't need the velocities at the faces or corners at all. The whole point of the scheme is that you compute the v x B terms at cell centers and then extrapolate that quantity to the corners. When you do that, the velocity by itself never appears at the corners.
> > 
> > 
> > I agree that is a point of the scheme, but to me the more salient feature is that Balsara's Riemann solver treats supersonic states whereas the LD04 Riemann solver assumes you are always in the subsonic strongly interacting state. So for low order reconstructions, the Balsara Riemann solver should be significantly less numerically resistive.
> 
> OK, that's fine, but in that case I think your taxonomy of possible combinations is a bit confusing. It seems like there are two independent axes that could be varied. One is how you reconstruct the corner EMFs, for which we have three possibilities:
> 
> 1. Felker & Stone scheme as originally published
> 2. Our variant of Felker & Stone with velocities coming from the Riemann solver.
> 3. The Balsara scheme where you compute v x B at cell centers and then extrapolate that to corners.
> 
> Then we have two distinct Riemann solvers: the LD04 one and the Balsara one. So I guess in principle there are 6 distinct combinations. I don't know that it makes sense to test all of them, but from a code legibility standpoint we should probably separate out the flags that control the choice of reconstruction to the corner scheme and those that control the corner Riemann solver, since in principle you can (I think) mix and match any of these, though of course not all are guaranteed to work.

That makes sense. That's a better taxonomy.

lizmcole committed:
Merge branch 'development' into Balsara-computeemf

lizmcole commented:
@BenWibking I tested with reconstruction method 3, it was less stable than the regular method. 

I removed the extra variables.

lizmcole committed:
removed extra comments, changed hook, reduced ghost cells in computeemfbalsara

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

lizmcole committed:
Update QuokkaSimulation.hpp to correct default method

lizmcole committed:
Update mhd_system.hpp to fix errors

lizmcole committed:
Update mhd_system.hpp

lizmcole committed:
Merge branch 'development' into Balsara-computeemf

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

lizmcole committed:
fixed error in computeemf_Balsara call

AstroKriel commented on src/QuokkaSimulation.hpp:
```diff
@@ -169,7 +169,9 @@ template <typename problem_t> class QuokkaSimulation : public AMRSimulation<prob
 
 	EMFAvgType emfAveragingType_ = EMFAvgType::LD04; // method to use to average EMF at edges
-	int emf_scheme_ = 1; // 0 == Felker and Stone scheme state; 1 == Felker and Stone scheme using the FC velocity from Riemann solver (default)
+	int emf_scheme_ = 1;				 // 0 == Felker and Stone scheme state;
```
Are the test problems actually running on this new scheme in the GitHub workflows (in this PR)? I think not (correct me if I'm missing something), since the default emf_scheme is set to 1 (our amended Felker & Stone scheme), unless the input files explicitly set it to 2 (the number for this scheme). Could I suggest temporarily setting this to 2, by default, so we make sure the scheme you’ve implemented actually gets run in this PR?

lizmcole replied:
Done

AstroKriel commented on src/QuokkaSimulation.hpp:
```diff
@@ -169,7 +169,9 @@ template <typename problem_t> class QuokkaSimulation : public AMRSimulation<prob
 
 	EMFAvgType emfAveragingType_ = EMFAvgType::LD04; // method to use to average EMF at edges
-	int emf_scheme_ = 1; // 0 == Felker and Stone scheme state; 1 == Felker and Stone scheme using the FC velocity from Riemann solver (default)
+	int emf_scheme_ = 1;				 // 0 == Felker and Stone scheme state;
```
Mhmm, just requesting this change again, now that I've added myself as a reviewer. I do think this should be done to make sure all of the automated tests work.

lizmcole committed:
Changed default emf scheme to balsara

BenWibking commented:
@claude Update the documentation with the new MHD runtime parameters.

claude[bot] committed:
docs: add mhd.emf_scheme parameter to documentation

- Added documentation for the new emf_scheme parameter
- Describes all three available EMF computation schemes
- Notes default value is 2 (Balsara 2025 method)

Co-authored-by: Ben Wibking <BenWibking@users.noreply.github.com>

BenWibking commented:
@claude Modify the `mhd.emf_scheme` runtime parameter so it accepts a string value using `AMREX_ENUM`, rather than using magic numbers to specify the three allowable options. Also, update the documentation accordingly.

claude[bot] committed:
refactor: use AMREX_ENUM for emf_scheme parameter

- Replace integer emf_scheme parameter with EMFScheme enum
- Update parameter type from int to EMFScheme throughout codebase
- Use descriptive names: FelkerStone, FelkerStoneFaceVel, Balsara2025
- Update documentation to show string values for the parameter
- Default remains Balsara2025 as before

Co-authored-by: Ben Wibking <BenWibking@users.noreply.github.com>

BenWibking commented:
/azp run

markkrumholz commented:
@BenWibking, I'd suggest holding off on this PR, because we're going to change these parameters shortly when we refactor to separate the parameter controlling the extrapolation scheme and the ones controlling the 2D Riemann solver.

markkrumholz commented:
Sorry, just to clarify: I meant that we shouldn't bother updating the input file to select the scheme, not the whole PR.

lizmcole committed:
Merge branch 'development' into Balsara-computeemf

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

BenWibking commented:
Code looks good.

What is the longer term plan for testing all of the combinations of schemes in the CI?

markkrumholz commented:
> Code looks good.
> 
> What is the longer term plan for testing all of the combinations of schemes in the CI?

I don't know that we necessarily want to test all, because not all of them are necessarily going to play well together -- we don't know yet. I would therefore advocate that we first get in the PR where we refactor the code to let us control the extrapolation and Riemann solver independently, do some preliminary testing to see which combinations are viable, and then set up the Alfven wave and fast wave tests for the combinations that we decide are viable. We might also then do a couple of the larger test problems for one or two recommended combinations on top of that.

BenWibking commented:
> > Code looks good.
> > What is the longer term plan for testing all of the combinations of schemes in the CI?
> 
> I don't know that we necessarily want to test all, because not all of them are necessarily going to play well together -- we don't know yet. I would therefore advocate that we first get in the PR where we refactor the code to let us control the extrapolation and Riemann solver independently, do some preliminary testing to see which combinations are viable, and then set up the Alfven wave and fast wave tests for the combinations that we decide are viable. We might also then do a couple of the larger test problems for one or two recommended combinations on top of that.

Ok. My only comment is that it would be useful to disable or remove the combinations that we find don't really work. That way there's no risk of users accidentally turning on something that they might be expecting to work.

markkrumholz commented:
> > > Code looks good.
> > > What is the longer term plan for testing all of the combinations of schemes in the CI?
> > 
> > 
> > I don't know that we necessarily want to test all, because not all of them are necessarily going to play well together -- we don't know yet. I would therefore advocate that we first get in the PR where we refactor the code to let us control the extrapolation and Riemann solver independently, do some preliminary testing to see which combinations are viable, and then set up the Alfven wave and fast wave tests for the combinations that we decide are viable. We might also then do a couple of the larger test problems for one or two recommended combinations on top of that.
> 
> Ok. My only comment is that it would be useful to disable or remove the combinations that we find don't really work. That way there's no risk of users accidentally turning on something that they might be expecting to work.

Agree. I think we won't know until we've done the testing.

lizmcole committed:
Merge branch 'development' into Balsara-computeemf

lizmcole committed:
Merge branch 'development' into Balsara-computeemf

lizmcole commented:
Any other changes I have forgotten to make? Everything seems to be working.

markkrumholz commented:
Looks good to me. If you're happy, you can merge as soon as the automated test suite passes.

chongchonghe commented:
The AlfvenWaveCircular test failed due to a nan error:

```
[FOFC-1] flux correcting 80 cells on level 0
Coordinates: (0.2734375, 0.5625, 0.0625):  At cell (17,4,0) in Box ((0,0,0) (63,7,7) (0,0,0)): -nan, nan, nan, nan, -nan, -nan
[FOFC-1] Flux correction failed:
Coordinates: (0.2734375, 0.5625, 0.0625):  At cell (17,4,0) in Box ((0,0,0) (63,7,7) (0,0,0)): -nan, -nan, -nan, -nan, -nan, nan
[FOFC-1] failed for 80 cells on level 0
	>> WARNING: Hydro advance failed on level 0

QUOKKA FATAL ERROR
Hydro update exceeded max_retries on level 0. Cannot continue, crashing...
```

@lizmcole can you take a look?
